peas/tests/serial_handlers/protocol_arduino_simulator.py

```diff
@@ -0,0 +1,422 @@
+            self.read_and_handle_relay_queue()
+            if self.pending_json_bytes and self.next_chunk_time < self.next_message_time:
+                next_time = self.next_chunk_time
+            else:
+                next_time = self.next_message_time
+            self.read_relay_queue_until(next_time)
+
+    def read_and_handle_relay_queue(self):
+        """Read all available entries in the relay queue, then handle any complete commands."""
+        while not self.relay_queue.empty():
+            b = self.relay_queue.get_nowait()
+            assert isinstance(b, (bytes, bytearray))
+            self.pending_relay_bytes.extend(b)
+        self.handle_pending_relay_bytes()
+
+    def handle_pending_relay_bytes(self):
+        """Process complete relay commands. NOT YET IMPLEMENTED."""
+        if len(self.pending_relay_bytes):
+            self.logger.error('ArduinoSimulator.handle_pending_relay_bytes NOT YET IMPLEMENTED')
+            self.logger.error('Dropping bytes: {!r}', self.pending_relay_bytes)
+            self.pending_relay_bytes.clear()
+        pass
+
+    def read_relay_queue_until(self, next_time):
+        """Wait for a little bit for relay input, not more than the time until the next action."""
+        now = datetime.datetime.now()
+        if now >= next_time:
+            # Already reached the time for the next main loop event,
+            # so return to repeat the main loop.
+            return
+        remaining = (next_time - now).total_seconds()
+        assert remaining > 0
+        self.logger.info('ArduinoSimulator.read_relay_queue_until remaining={}', remaining)
+        try:
+            b = self.relay_queue.get(block=True, timeout=remaining)
+        except queue.Empty:
+            return
+        if b:
+            assert isinstance(b, (bytes, bytearray))
+            self.pending_relay_bytes.extend(b)
+            self.handle_pending_relay_bytes()
+
+    def output_next_chunk(self, now):
+        """Output one chunk of pending json bytes."""
+        self.next_chunk_time = now + self.chunk_delta
+        if len(self.pending_json_bytes) == 0:
+            return
+        l = min(self.chunk_size, len(self.pending_json_bytes))
+        chunk = bytes(self.pending_json_bytes[0:l])
+        del self.pending_json_bytes[0:l]
+        if self.json_queue.full():
+            self.logger.info('Dropping chunk because the queue is full')
+            return
+        self.json_queue.put_nowait(chunk)
+        self.logger.debug('output_next_chunk -> {}', chunk)
+
+    def generate_next_message(self, now):
+        """Append the next message to the pending bytearray and scheduled the next message."""
+        b = self.generate_next_message_bytes(now)
+        self.pending_json_bytes.extend(b)
+        self.next_message_time = datetime.datetime.now() + self.message_delta
+
+    def generate_next_message_bytes(self, now):
+        """Generate the next message (report) from the simulated Arduino."""
```

Comment:
Not necessary for now but might be nice to change other values besides `millis` and `report_num`.

Reply:
I figured that would be gilding the lily. And yes, I thought of it. 😁